### ISSUE TITLE
The wrong turn is announced in voice navigation and displayed in the widget.

### ISSUE THREAD

**Issue by anastasiia936** [osmandapp/OsmAnd#11991]
Following the navigation, I need to go straight, but in the widgets, the arrows indicate that I need to turn left.

![Screenshot_20210622-174651](https://user-images.githubusercontent.com/80623554/123260583-6539ca80-d4fe-11eb-8b67-1d24ba8b1c66.png)

**Comment by woheller69** [comment]
I just had the same (?) problem. I think navigation was one waypoint ahead. So it did not tell me what to do next but what to do at the waypoint after.
Stopped navigation and started again -> still wrong.
Terminated app and started navigation again -> worked

F-Droid version 3.9.10

**Comment by Martinelli1978** [comment]
I am having this problem too, but then as a pedastrian.  For example yesterday I had this gpx and was repeatedly told to go left (or right) when I actually had to go in the opposite direction. This is not the first times this has happened, with multiple gpx regardless if I have drawn them myself or if they are premade.  It is very annoying because I still have to walk while looking at the map while it would be nice to just rely on the speech. Yesterday I made a screenshot where I was told to go left, while that wasn't even a possibilty. 
![Screenshot_2021-09-14-11-40-30-247_net osmand](https://user-images.githubusercontent.com/90757989/133403307-f5edd52e-3c64-4bfa-afdf-6d200efc7c38.jpg)

**Comment by woheller69** [comment]
> I just had the same (?) problem. I think navigation was one waypoint ahead. So it did not tell me what to do next but what to do at the waypoint after. Stopped navigation and started again -> still wrong. Terminated app and started navigation again -> worked
> 
> I used bicycle mode. In pedestrian mode I never noticed that problem.
> 
> F-Droid version 3.9.10

In the meantime I also had the problem in pedestrian mode.

**Comment by woheller69** [comment]
I do not think these changes will fix my problem because my problem is not one with several lanes.
For biking or hiking I usually import a GPX or use the tool to plan a route.
Then I navigate along this route and after a while (maybe somewhere in the forest, with no other lanes) wrong navigation instructions are given. 
If I stop navigation and start again so the route is calculated again everything, incuding instructions for current position, is fine - at least for a while.
Then the problem may occur again.
In my case that usually happens at positions where nothing has to be done and the distance to AND the direction of the next turn are wrong.

**Comment by vshcherb** [comment]
Yes it should be separate issue, this one is about turn lanes tagging and parsing, yours is about turn detection itself
